aptos-move/mvhashmap/src/versioned_group_data.rs

```diff
@@ -0,0 +1,667 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::types::{Flag, Incarnation, MVGroupError, ShiftedTxnIndex, TxnIndex, Version};
+use aptos_types::write_set::TransactionWrite;
+use claims::assert_some;
+use crossbeam::utils::CachePadded;
+use dashmap::DashMap;
+use serde::Serialize;
+use std::{
+    collections::{btree_map::BTreeMap, HashMap, HashSet},
+    fmt::Debug,
+    hash::Hash,
+    sync::Arc,
+};
+
+struct GroupEntry<V> {
+    // Note: can be a raw pointer (different data-structure holds the value during the
+    // lifetime), but would require unsafe access.
+    value: Arc<V>,
+    incarnation: Incarnation,
+    flag: Flag,
+}
+
+impl<V: TransactionWrite> GroupEntry<V> {
+    fn new(value: Arc<V>, incarnation: Incarnation) -> Self {
+        Self {
+            value,
+            incarnation,
+            flag: Flag::Done,
+        }
+    }
+}
+
+/// Represents a group value, i.e. a key that does not correspond to a single value,
+/// but instead a collection of values each associated with a tag.
+///
+/// Implementation note: due to DashMap in VersionedGroupData, the updates are atomic.
+/// If this changes, we must maintain invariants on insertion / deletion order among
+/// members (e.g. versioned_map then idx_to_update then member_tags, deletion opposite).
+pub(crate) struct VersionedGroupValue<T, V> {
+    /// While versioned_map maps tags to versioned entries for the tag, idx_to_update
+    /// maps a transaction index to all corresponding group updates. ShiftedTxnIndex is used
+    /// to dedicated index 0 for base (storage version, prior to block execution) values.
+    versioned_map: HashMap<T, BTreeMap<ShiftedTxnIndex, CachePadded<GroupEntry<V>>>>,
+    /// Note: if we do not garbage collect final idx_to_update contents until the end of
+    /// block execution (lifetime of the data-structure), then we can have other structures
+    /// hold raw pointers to the values as an optimization.
+    idx_to_update: BTreeMap<ShiftedTxnIndex, CachePadded<HashMap<T, Arc<V>>>>,
+
+    /// Group membership is addressed in the AIP:
+    /// https://github.com/aptos-foundation/AIPs/blob/main/aips/aip-9.md
+    ///
+    /// In summary, existing resources may not be added to a group, and existing members
+    /// may not be removed from a group. However, a newly created resource may become
+    /// a member, and also only members whose corresponding instances exist will be stored
+    /// in a given group at a given time.
+    ///
+    /// We maintain the list of all tags the key has been associated with (either
+    /// in storage, or in any of the writes), and also use it to determine the 'group_size'
+    /// when requested. Group size is computed as the sum of sizes (length of raw bytes)
+    /// of all latest inner values of all member tags, plus serialized length of tags.
+    member_tags: HashSet<T>,
```

Comment:
Why do we need these, isn't it always equal to the key set of versioned_map?

Reply:
oh my god, good point :D :D thanks

Reply:
I think when I did this, I hadn't combined base and updated values with shiftedindex and forcing initialization, and then I never noticed.